Add photo gallery

tiembo opened the pull request.

tiembo committed:
Update build files to fetch Unsplash API keys

tiembo committed:
Add retrofit dependencies, update lifecycle version

tiembo committed:
Add the Unsplash Service

tiembo committed:
Add Unsplash data classes

tiembo committed:
Add okhttp Logging interceptor

tiembo committed:
Add Unsplash respository, search response, and search result

tiembo committed:
Checkpoint: pics load from network

tiembo committed:
Remove unused UNSPLASH_SECRET_KEY build config field

tiembo committed:
Add KDoc for data classes and remove unused fields

tiembo committed:
Add photo attribution

Display the user's name associated with each image

tiembo committed:
Use plant name for search instead of temporary placeholder

tiembo committed:
Add toolbar to gallery screen

tiembo committed:
Merge branch 'master' into gallery

tiembo committed:
Cleanup code

florina-muntenescu commented on app/src/main/java/com/google/samples/apps/sunflower/PlantDetailFragment.kt:
```diff
@@ -105,12 +105,23 @@ class PlantDetailFragment : Fragment() {
+
+            // TODO: can !! be avoided here?
+            plantDetailName.setOnClickListener {
+                navigateToGallery(plantDetailViewModel.plant.value!!.name, view!!)
```
i'd do
```plantDetailViewModel.plant.value?.let { ... } ```

tiembo replied:
Done

tiembo committed:
Navigate to gallery using button

Button is hidden if the Unsplash API key is not set

Also rename FAB binding adapter from `isGone` to `isFabGone` to avoid overlap with View binding adapter that's named `isGone`

tiembo committed:
Use icon instead of button to navigate to gallery

tiembo committed:
Link to Unsplash photo user's profile page

tiembo committed:
Move icon so that the FAB does not overlap it

tiembo committed:
Add placeholder strings for non-English languages

tiembo committed:
Add instructions for obtaining an Unsplash API key

florina-muntenescu commented on app/src/main/java/com/google/samples/apps/sunflower/adapters/GalleryAdapter.kt:
```diff
@@ -0,0 +1,77 @@
+ * Adapter for the [RecyclerView] in [GalleryFragment].
+ */
+
+class GalleryAdapter : ListAdapter<UnsplashPhoto, ViewHolder>(GalleryDiffCallback()) {
```
You should be able to use `GalleryViewHolder` here instead of `ViewHolder`. Like this you get rid of the `holder as GalleryViewHolder` below.

tiembo replied:
Done

manuelvicnt commented on app/src/main/java/com/google/samples/apps/sunflower/GalleryFragment.kt:
```diff
@@ -0,0 +1,88 @@
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.FlowPreview
+
+@ExperimentalCoroutinesApi
```
Nit: you can avoid annotating the classes or methods with a workaround in the gradle file. Check Tivi's for example:
https://github.com/chrisbanes/tivi/blob/main/build.gradle#L98

tiembo replied:
Nice! Added

manuelvicnt commented on app/src/main/java/com/google/samples/apps/sunflower/data/UnsplashRepository.kt:
```diff
@@ -0,0 +1,50 @@
+@ExperimentalCoroutinesApi
+class UnsplashRepository (private val service: UnsplashService) {
+
+    private val searchResult = ConflatedBroadcastChannel<UnsplashSearchResult>()
```
Maybe use `MutableStateFlow` here? More efficient than this

tiembo replied:
Will update this in a separate PR

manuelvicnt commented on app/src/main/java/com/google/samples/apps/sunflower/viewmodels/GalleryViewModel.kt:
```diff
@@ -0,0 +1,50 @@
+    val repoResult: LiveData<UnsplashSearchResult> = queryLiveData.switchMap { queryString ->
+        liveData {
+            val repos = repository.getSearchResultStream(queryString).asLiveData(Dispatchers.Main)
+            emitSource(repos)
```
Do you need `Dispatchers.Main` here? I think `asLiveData()` should be enough

tiembo replied:
Done

manuelvicnt commented on app/src/main/java/com/google/samples/apps/sunflower/viewmodels/PlantDetailViewModel.kt:
```diff
@@ -40,4 +41,6 @@ class PlantDetailViewModel(
         }
     }
+
+    fun hasValidUnspliashKey() = (BuildConfig.UNSPLASH_ACCESS_KEY != "null")
```
Typo
```suggestion
    fun hasValidUnsplashKey() = (BuildConfig.UNSPLASH_ACCESS_KEY != "null")
```

tiembo replied:
Done

tiembo committed:
Use GalleryViewHolder instead of ViewHolder in adapter

tiembo committed:
Fix typo

tiembo committed:
Enable Coroutines and Flow APIs at the project level

This also allows removal of individual class and method annotations as the project level setting serves the same purpose.

tiembo committed:
Don't specify coroutine context

Use default one instead

tiembo commented:
Merged to android/sunflower via another PR.
